src/main/java/tel/schich/javacan/BcmMessage.java

```diff
@@ -0,0 +1,363 @@
+     */
+    public int getCount() {
+        return buffer.getInt(base + OFFSET_COUNT);
+    }
+
+    /**
+     * The {@code interval1} has different meanings depending on the {@link #getOpcode()} of the
+     * message:
+     * <ul>
+     * <li><b>When used with {@link BcmOpcode#TX_SETUP}:</b></br>
+     * The broadcast manager sends {@link #getCount()} messages with this interval, then continue to
+     * send at {@link #getInterval2()}. If only one timer is needed set
+     * {@link BcmMessageBuilder#count(int) count} to {@code 0} and {@link BcmMessageBuilder#ival1
+     * interval1} to {@code null}.</li>
+     * <li><b>When used with {@link BcmOpcode#RX_SETUP}:</b></br>
+     * Send {@link BcmOpcode#RX_TIMEOUT} when a received message is not received again within the given
+     * interval. When {@link BcmFlag#STARTTIMER} is set, the timeout detection is activated directly -
+     * even without a former CAN frame reception.</li>
+     * </ul>
+     *
+     * @return the duration or {@code null} if it is not set
+     */
+    public Duration getInterval1() {
+        long sec = getPlatformLong(base + OFFSET_IVAL1_TV_SEC);
+        long usec = getPlatformLong(base + OFFSET_IVAL1_TV_USEC);
+        return (sec + usec) != 0 ? Duration.ofSeconds(sec).plusMillis(usec) : null;
+    }
+
+    /**
+     * The {@code interval2} has different meanings depending on the {@link #getOpcode()} of the
+     * message:
+     * <ul>
+     * <li><b>When used with {@link BcmOpcode#TX_SETUP}:</b></br>
+     * see {@link #getInterval1()}</li>
+     * <li><b>When used with {@link BcmOpcode#RX_SETUP}:</b></br>
+     * Throttle the received message rate down to the value of {@code interval2}. This is useful to
+     * reduce messages for the application when the signal inside the CAN frame is stateless as state
+     * changes within the {@code interval2} duration may get lost.</li>
+     * </ul>
+     *
+     * @return the duration or {@code null} if it is not set
+     */
+    public Duration getInterval2() {
+        long sec = getPlatformLong(base + OFFSET_IVAL2_TV_SEC);
+        long usec = getPlatformLong(base + OFFSET_IVAL2_TV_USEC);
+        return (sec + usec) != 0 ? Duration.ofSeconds(sec).plusMillis(usec) : null;
+    }
+
+    /**
+     * Returns the CAN ID of this message.
+     */
+    public int getCanId() {
+        return buffer.getInt(base + OFFSET_CAN_ID);
+    }
+
+    /**
+     * Returns the number of frames in this message.
+     */
+    public int getFrameCount() {
+        return buffer.getInt(base + OFFSET_NFRAMES);
+    }
+
+    /**
+     * Returns a single frame of this message.
+     *
+     * @param index of the frame; ({@code 0 <= index < frameCount})
+     * @return the frame
+     * @throws IllegalArgumentException if the message buffer contains no frame for that index
+     */
+    public CanFrame getFrame(int index) {
+        ByteBuffer frameBuffer = buffer.duplicate();
```

Comment:
It probably doesn't make a huge difference, except maybe possible race-conditions, but have you considered reusing the internal buffer directly?

Reply:
Yes i did consider it, but it seemed more complicated than using an own buffer instance.
The `duplicate` buffer has the advantage that i have independent position and limit without any real memory duplication. It's just the cost of the object creation.

So i can modify the limit and position of the `frameBuffer` without worrying about restoring it in the original buffer object after i used it.

Reply:
> That whole concept of having the ByteBuffer as value holder looks strange to me. As it effectively prevents to make the BcmMessage object immutable. As soon as the buffer gets modified the object is modified. For me it's a bit of a code smell.

I completely agree here. You could consider the the CanFrame to be more like an immutable view into a piece of mutable memory than an actual. That's how it was intended to be used and how I use it in libs/apps on top. And from that angle the BcmMessage could be considered a view with multiple subviews into subsections of the same memory.

But back to the question: I agree that duplicating the ByteBuffer object for independent positions is probably a good idea with basically no disadvantage. Since all access methods reset the position to where they need it, resetting the position would not be your concern as a user, but as soon as multiple threads get involved this gets hairy.

Reply:
I don't quite understand your point with multi-threading.
The `ByteBuffer` object created by the `duplicate` method share the same memory region with the original buffer. The other attributes are independent.
So if you **don't create a duplicate** and operate on this buffer object from multiple threads then you might have an issue at hand, as one thread may modify the position, limit, a.s.o. while another is trying to read. This may lead to false reading.
With the duplicate buffer this side effect cannot happen.

A whole other thing would be any write access to the buffer. This affects all objects, also the duplicates.
As you pointed out, the `BcmMessage` and `CanFrame` should be considered as views onto a memory region. So i think that the `duplicate` buffer is a good idea, as it IMHO improves thread safety.

Reply:
My wording was a bad, I actually meant that *not* duplicating the buffer would risk race conditions. So I agree.